[SVCS-52] Adding support for mutli-page tiff files.

AddisonSchiller opened the pull request:
adding more to pdf renderer
tiff files are now rendered by the pdf renderer, and the pdf extension now has an exporter to handle tif/tiff files. Tiff files are now converted to pdfs for viewing

refs: https://openscience.atlassian.net/browse/SVCS-52

## Purpose
Add support for mutlipage tiff files.

## Summary of changes
Tiff files are sort of like pdfs, and can be converted to them. 
The pdf extension now handles tiff files. The pdf extension now has an exporter for tiff files
Reportlab has been added as a dependency. 
Some multipage tiffs will not render correctly. see testing notes. 

## Testing QA notes:
below is a copy and paste of my testing notes on the jira ticket. you can find the zip file over there

Testing Notes:

Attached to the jira ticket is a zip containing some tif/tiff files to use for testing.

Due to the aspects of what this commit is doing, the following should also be tested:
    Other pdf files. Try a few regular pdf files and make sure they render properly.
    Any files rendered by the unoconv renderer. (ie docx) The unoconv renderer ships files to the pdf renderer.

cslzchen commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+            except EOFError:
+                break
+
+            # Center the image and draw it in the canvas
```
Be consistent on comments style (a few occurrences).
* should we capitalize the first letter of the first word (whichever you prefer)
* should we add punctuation at the end of the line (usually not)

cslzchen commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+
+            self.tiff_to_pdf(image, max_size)
+            image.close()
+
```
Remove this empty line.

cslzchen commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+                original_exception=err,
+                code=400,
+            )
+        except(ValueError) as err:
```
```python
except ValueError as err:
```

AddisonSchiller replied:
as a note, the `ValueError` case is a bug with our current version of pillow having trouble opening certain formats of `.tiff`. Once pillow is updated to 4.3 or higher, this last error section can go away.

cslzchen commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+            self.tiff_to_pdf(image, max_size)
+            image.close()
+
+        except (UnicodeDecodeError, IOError) as err:
```
The code and exception message looks similar. Why not combining the two `except` in one? (A question not a suggestion.)

AddisonSchiller replied:
The `ValueError` case is a bug with our current version of pillow having trouble opening certain formats of `.tiff`. Once pillow is updated to 4.3 or higher, this last error section can go away. (tested now and this is indeed the case. `broken.tiff` will open on pillow 4.3, but not 2.8.3 that we currently have. This is why the errors were separate. The tiff given was valid, just pillow messing up.

cslzchen commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+                export_format=type,
+                detected_format=imghdr.what(self.source_file_path),
+                original_exception=err,
+                code=400,
```
Given this is a new file, let's use `HTTPStatus` ([LINK](https://docs.python.org/3/library/http.html#http-status-codes)).
```python
from http import HTTPStatus
```
```python
code=HTTPStatus.BAD_REQUEST
```

cslzchen commented on mfr/extensions/pdf/render.py:
```diff
@@ -13,7 +15,19 @@ class PdfRenderer(extension.BaseRenderer):
+        elif settings.EXPORT_TYPE:
+            exported_url.args['format'] = settings.EXPORT_TYPE
+        else:
+            return self.TEMPLATE.render(base=self.assets_url, url=self.metadata.download_url)
```
What's the conditions for reaching this branch? Is this unexpected?

cslzchen commented on mfr/extensions/pdf/render.py:
```diff
@@ -13,7 +15,19 @@ class PdfRenderer(extension.BaseRenderer):
+            return self.TEMPLATE.render(base=self.assets_url, url=self.metadata.download_url)
+
+        exported_url = furl.furl(self.export_url)
+        if settings.EXPORT_MAXIMUM_SIZE and settings.EXPORT_TYPE:
```
This reduce the # of `bool` checks from 3 to 2.
```python
if settings.EXPORT_TYPE:
    if settings.EXPORT_MAXIMUM_SIZE
        exported_url.args['format'] = '{}.{}'.format(settings.EXPORT_MAXIMUM_SIZE, settings.EXPORT_TYPE)
    else:
        exported_url.args['format'] = settings.EXPORT_TYPE
    self.metrics.add('needs_export', True)
    return self.TEMPLATE.render(base=self.assets_url, url=exported_url.url)

# TODO: is this an unexpected case?
return self.TEMPLATE.render(base=self.assets_url, url=self.metadata.download_url)
```

AddisonSchiller replied:
This was mostly copied from the image renderer. I assume that the TODO case here is when the settings are not properly defined. Going with your approach here but leaving the unexpected case in for completion etc.

cslzchen replied:
Agreed. Similarly, update image renderer if possible.

cslzchen commented on mfr/extensions/pdf/exceptions.py:
```diff
@@ -0,0 +1,23 @@
+from mfr.core.exceptions import ExporterError
+
+
+class PillowImageError(ExporterError):
```
Seems duplicated, but I guess you have a good reason for not using `mfr.extensions.image.exceptions.PillowImageError`?

AddisonSchiller replied:
It is duplicated, but it doesnt make sense to import an exception from a different extension (for the most part, extensions dont interact much) thats why its copied this way. Thoughts?

cslzchen replied:
I see what you mean and I agree. Let's leave this here for @felliott in Phase 2 Review.

felliott replied:
I'm okay with duplicating this, but the docstrings and `__TYPE` should be consistent with the new location.  `__TYPE` should probably be `pdf_pillow`.

AddisonSchiller commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+
+    def export(self):
+        parts = self.format.split('.')
+        type = parts[-1].lower()
```
This code is directly from the image  exporter.

cslzchen replied:
Let's change this here. Using reserved word is a huge issue. If possible (not causing too much trouble), update the image exporter as well. Thanks.

AddisonSchiller replied:
I will update the image exporter on the ticket I just did for it. 👍

AddisonSchiller commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+    def tiff_to_pdf(self, tiff_img, max_size):
+        max_pages = 40
+        width, height = tiff_img.size
+        c = canvas.Canvas(self.output_file_path)
```
There should be no file at output_file_path. Thats the path MFR wants the exported file to be saved to.

cslzchen replied:
Ah, I see. I thought there is a temp file created.

AddisonSchiller commented on mfr/extensions/pdf/__init__.py:
```diff
@@ -1 +1,2 @@
 from .render import PdfRenderer  # noqa
+from .export import PdfExporter # noqa
```
Leaving this like we talked about on other issue

AddisonSchiller commented on mfr/extensions/pdf/export.py:
```diff
@@ -0,0 +1,85 @@
+from reportlab.pdfgen import canvas
+
+from mfr.core import extension
+from mfr.extensions.pdf import exceptions
```
Imghdr is a base lib, so leaving it with os, moving others.

felliott commented on mfr/extensions/pdf/render.py:
```diff
@@ -13,6 +15,20 @@ class PdfRenderer(extension.BaseRenderer):
+        if settings.EXPORT_TYPE:
+            if settings.EXPORT_MAXIMUM_SIZE:
+                exported_url.args['format'] = '{}.{}'.format(settings.EXPORT_MAXIMUM_SIZE,
+                                                                    settings.EXPORT_TYPE)
```
nitpick: indentation

felliott commented on mfr/extensions/pdf/settings.py:
```diff
@@ -0,0 +1,9 @@
+EXPORT_TYPE = config.get('EXPORT_TYPE', 'pdf')
+EXPORT_MAXIMUM_SIZE = config.get('EXPORT_MAXIMUM_SIZE', '1200x1200')
+EXPORT_EXCLUSIONS = config.get('EXPORT_EXCLUSIONS', ['.pdf', ])
+EXPORT_MAX_PAGES = config.get('EXPORT_MAX_PAGES', 40)
```
The above two config settings will have problems if set from the env.  Envvars are always strings; they won't be ints or lists.  I believe there are some helper methods in the config base class to cast these properly.

felliott commented on tests/extensions/pdf/test_exporter.py:
```diff
@@ -0,0 +1,93 @@
+
+from mfr.extensions.pdf import settings
+from mfr.extensions.pdf import exceptions
+from mfr.extensions.pdf import PdfExporter
```
The above could be collapsed.

felliott commented on tests/extensions/pdf/test_exporter.py:
```diff
@@ -0,0 +1,93 @@
+
+class TestPdfExporter:
+    '''
+    Opening and veryifying pdfs with report lab is a paid feature
```
Nitpick: "verifying".  Also, "report lab" could probably use a little more context here, it's only mentioned/used in requirements.txt and the renderer file.

AddisonSchiller replied:
reworded/fixed typos/added new typos
complete

felliott commented on tests/extensions/pdf/test_exporter.py:
```diff
@@ -0,0 +1,93 @@
+    '''
+    Opening and veryifying pdfs with report lab is a paid feature
+    so these tests are mostly just to see if the file gets created
+    in the correct place
```
Is there anything else in the stdlib that can assert that the output is actually a pdf?  If not, no worries, it would just be good to assert that a pdf is produced, rather than e.g. and html error message.

felliott replied:
Followup: it looks like PDFs have a signature: https://en.wikipedia.org/wiki/List_of_file_signatures

AddisonSchiller replied:
This is great, implementing

AddisonSchiller replied:
I found some libraries that use system libraries to check things (libmagic used by python-magic) etc. They would require adding things to the docker file or something like that. I just went with opening the file, and pulling the signature out directly and looking at it. Slightly crude but works fine.

felliott replied:
:+1:

felliott commented on tests/extensions/pdf/test_exporter.py:
```diff
@@ -0,0 +1,93 @@
+        assert os.path.exists(directory)
+
+        exporter.export()
+        assert os.path.exists(output_file_path)
```
It looks like the successful tests and the failing tests could be parameterized, since we don't have the ability to do deep examination of the pdfs.

AddisonSchiller replied:
What do you mean by this?

felliott replied:
The three passing tests are all very similar, they just take slightly different arguments.  You could parameterize them a la [this WB s3 test](https://github.com/CenterForOpenScience/waterbutler/blob/develop/tests/providers/s3/test_provider.py#L119).  That way if you update one later, the others get updated, too.

AddisonSchiller replied:
This is great,
completed.

AddisonSchiller commented on mfr/extensions/pdf/settings.py:
```diff
@@ -0,0 +1,9 @@
+
+EXPORT_TYPE = config.get('EXPORT_TYPE', 'pdf')
+EXPORT_MAXIMUM_SIZE = config.get('EXPORT_MAXIMUM_SIZE', '1200x1200')
+EXPORT_EXCLUSIONS = config.get('EXPORT_EXCLUSIONS', ['.pdf', ])
```
the image settings is done the same way (its mostly c/p from the image settings). I don't see a method in mfr.settings to handle lists. 
Should I be feeding this a string and making the list myself in the renderer or something like that?

also, should the image extension be changed as well?

AddisonSchiller replied:
Other small note: even though the list is turned into a string, it will still find '.pdf' in that string and exclude it. Not sure if its intended to work that way, but that must be why its working for images settings.  (this is why it works even though it seems like it shouldn't)

As mentioned above im open to any fix for this

felliott replied:
Blarf, sorry I was misremembering. The similar example I was thinking of was [here](https://github.com/CenterForOpenScience/modular-file-renderer/blob/develop/mfr/server/settings.py#L35).  We support passing a list via the env by requiring it to be a space-separated string that gets split upon construction.  And yes, the image renderer settings should be doing it this way, but we never bothered to audit the settings when we set it up.  There are probably lots of settings that need this, so we'll create a separate janitorial ticket to audit and clean these up.

AddisonSchiller replied:
Done, and ticket made.

AddisonSchiller commented on mfr/extensions/pdf/settings.py:
```diff
@@ -0,0 +1,9 @@
+EXPORT_TYPE = config.get('EXPORT_TYPE', 'pdf')
+EXPORT_MAXIMUM_SIZE = config.get('EXPORT_MAXIMUM_SIZE', '1200x1200')
+EXPORT_EXCLUSIONS = config.get('EXPORT_EXCLUSIONS', ['.pdf', ])
+EXPORT_MAX_PAGES = config.get('EXPORT_MAX_PAGES', 40)
```
complete

AddisonSchiller commented on mfr/extensions/pdf/exceptions.py:
```diff
@@ -0,0 +1,23 @@
+    and relating to the Pillow Library should inherit from PillowImageError
+    """
+
+    __TYPE = 'image_pillow'
```
complete

AddisonSchiller commented on tests/extensions/pdf/test_exporter.py:
```diff
@@ -0,0 +1,93 @@
+
+from mfr.extensions.pdf import settings
+from mfr.extensions.pdf import exceptions
+from mfr.extensions.pdf import PdfExporter
```
complete

AddisonSchiller commented on tests/extensions/pdf/test_renderer.py:
```diff
@@ -1,26 +1,41 @@
 def metadata():
-    return ProviderMetadata('test', '.pdf', 'text/plain', '1234', 'http://wb.osf.io/file/test.pdf?token=1234')
+    return ProviderMetadata('test', '.pdf', 'text/plain', '1234',
+                        'http://wb.osf.io/file/test.pdf?token=1234')
```
complete

AddisonSchiller commented:
@felliott 
Random comments/CRR

Take a look at the settings questions, not quite sure how you want me to handle that.

On libtiff:
I took a look at it, and it was sort of inconclusive. All i could find is there seem to be some tiff files that label things not to the tiff standard. using libtiff might help with those? I don't think we are using it in our docker container atm (it might come auto installed though.)

I did try to install it and see if it changed the way any of my test tiff files rendered and it did not seem to. 

Just couldn't find enough information about it. Some of the docs linked to other libraries i originally looked at for rendering tiffs etc, and such. If you want i could definitely spend more time on this.

Also, leaving in additional development while I wait for responses to some of my questions

felliott commented:
I've worked a bunch with .tiff files in the past.  The tiff format is more of a container format than a image compression format.  It permits lots of weird formats and extensions and its a pretty popular format in science for that reason. I'm not usually a fan of adding dependencies speculatively, but I would make an exception for tiffs.  It would be good to add a comment explaining why it's in there.

AddisonSchiller commented:
Did some more research. It looks like libtiff5 comes default in Jessie. I read some things about the correct version of libtiff not being used. It shouldn't hurt to install python-libtiff and let pillow go crazy with it. So adding that to the PR as suggested.

AddisonSchiller commented:
Latest additions:
Confirmed (with @felliott ) that libtiff5-dev is working in the docker container
Added libtiff5-dev to the docker-file
Removed unused python libtiff thing

Turning on libtiff use in pillow fixed color issues on some tiffs
fixed resizing on large multipage tiffs (should now display all pages)
lots of refactoring/changes from first few versions.
Pillow is upgraded to 4.3.0 on this ticket. Standalone it will work with tiffs, but will break the image renderer. There is another ticket I have out that should be merged before this one that will fix what pillow 4.3.0 breaks in MFR.

felliott commented:
TerTIFFic!  Moving to RTM
